### DISCUSSION TITLE
Runtime router configuration

### DISCUSSION THREAD

**Discussion by marknefedov** [trillium-rs/trillium#90] [Ideas]
Is it possible with current implementation do something like this?
```rust
pub fn example(plugins: Arc<RwLock<Vec<PluginEndpoint>>>) -> Handler{
    let mut router = Router::new();
    for endpoint in plugins.read()... { // I don't remember exact syntax
        router = router.all(endpoint.path, endpoint.handler);
    }
    (router, super::not_found)
}
```
I noticed that most web-frameworks build route tree at compile/startup times so implementing runtime route updates is not really feasible. It would be cool to have ability to pass Conn to "handler" form wasm plugin that user installed.

**Comment by jbr** [comment]
Could you elaborate a bit more on the use case? I don't think I understand when the routes get updated. It should be possible though, regardless. Would incoming requests modify the route table?

**Reply by jbr** [reply]
Running untrusted wasm code would require a significant layer of insulation between the current trillium code and the untrusted code. It should be possible, but probably couldn't even directly interact with Conn.

However, yes it should be possible to modify the routes table at runtime. This would require holding the internal data inside of the Router inside of a RwLock or Mutex, which probably makes more sense to do in a fork of the current router, since most users shouldn't have to pay the cost of a lock on every request

Alternatively, perhaps it would be sufficient to `impl<H: Handler> Handler for RwLock<H>` and leave it up to the application to wrap the Router and make the `Arc<RwLock<Router>>` available on a Conn

**Comment by jalcine** [comment]
For what it's worth, I think this could be _indirectly_ done by using a wildcard path matcher and having application-code that does its own matching of a route against the plugins loaded here.

**Reply by jbr** [reply]
Exactly!

Routers can also be run in sequence — anything that isn't matched by the first one falls through to the next one, so in theory a "runtime-mutable router" could be run before or after some other router, even if the actual router implementations are different.

Meta note: One of the design goals of the current stage of trillium development is to support lots of different design patterns. Later on when the design space has been explored a bit more, maybe it'll make sense to offer guidance to people on stuff like whether it's preferable to do `(Router::new().get("/x", "x"), Router::new().get("/namespace/y", "y"))` vs `Router::new().get("/x", "x").get("/namespace/*", Router::new("/y", "y"))`

In particular, I'm especially interested in the use case of people publishing "sub-applications" as crates. Imagine like an OAuth2 endpoint that has its own routes, or an admin dashboard that shows realtime performance stats (including bundling frontend assets in the crate). It hopefully will be easy to mount those on apps without having to wire up each route
